Latest updates based on feedback. (#11)
Azure-Samples/azure-monitor-opencensus-python: azure_monitor/azfunc_sample/README.md

---
title: 'Sample: Configuring Function Apps with Python and OpenCensus'
description: Create Azure Functions using Python and OpenCensus.
author: givenscj
ms.author: givenscj
ms.devlang: python
ms.topic: sample
ms.date: 07/27/2022
ms.custom: []
---

> # Sample: Configuring Function Apps with Python and OpenCensus

Reviewer: ```suggestion
# Sample: Configuring Function Apps with OpenCensus SDK
```

> This is an example of how to add the various settings Python-based Azure Function applications to send dependencies calls to Azure Monitor.  You will run the samples both locally and deploy the Azure Functions to an Azure App Service resource you deployed as part of the setup script.

Reviewer: ```suggestion
This is an example of how to configure Azure Function applications to send dependencies calls to Azure Monitor.  You will run the samples both locally and deploy the Azure Functions to an Azure App Service resource you deployed as part of the setup script.
```

> ## 1 - Environment Setup

Reviewer: ```suggestion
## 1 - Setup OpenCensus SDK
```

> Follow the steps in the [setup documentation](/azure_monitor/readme.md).

Reviewer: ```suggestion
Follow the steps in the [documentation](/azure_monitor/readme.md) to setup OpenCensus SDK for you application.
```

> 

Reviewer: Is there a step needed here to link to functions docs for setup? I remember there is something needed to cause `func` to be used as a command below.

Reviewer: The target audience for this is a person who already knows what and how to set up an azure function app, which is why we need the link to their docs.

Author: Only setup needed is to download the azure functions core tools, and there is a link to that in the prereqs already.

## 2 - Utilize Function Apps with OpenCensus

> - Start the Azure Storage Emulator.

Reviewer: What is `Azure Storage Emulator` and how do you start it?

Reviewer: Oh I see that it is linked [here](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-emulator). Maybe we can mark this as `optional` since it is not really needed to get the sample running.

Author: One of the steps in the functions sample is to run the sample locally, and you would need it for that.  I added some extra text that it is for the function app sample.

- Switch to Visual Studio Code.
- Open a terminal window, run the following commands to start the Python functions locally:

```powershell
cd .\azure_monitor\azfunc_sample

python -m pip install -r requirements.txt

func start

> ```

Reviewer: Can we have a quick explanation of what the function app is doing so that users will know what to expect in their telemetry?

Reviewer: What does `MyHTTPFunction` do and `MyTimerFunction` do and how do they interact with each other and other components?

Author: Added details of the function apps.  I removed the timer function app as it is the same as the http one in terms of the core code.  Idea was to potentially have metrics with sampling for the timer, but we'll have to skip it.

- If you get an error about `psutil._psutil_windows` do the following:
  - Open the `.venv\Lib\site-packages` folder and delete the `psutil` and `psutil-5.9.1.dist-info` folders.
  - Then run the following:

    ```Python
    python -m pip install --upgrade pip
    python -m pip install psutil
    ```

- Browse to the HTTP function endpoint `http://localhost:7071/api/MyHttpFunction`.

> ## 3 - Deploy to Azure

Reviewer: Why is deploying to azure a step within our docs for monitoring? Is this needed in your opinion @mattmccleary ?

```powershell
cd .\azure_monitor\azfunc_sample

az login

func azure functionapp publish python-appinsights-SUFFIX   
```

- You should see the two function applications deployed successfully:

  ![The results of the function app publish command is displayed.](../media/python_functionapp_deploy.png "Review the results and ensure the function apps deployed succesfully.")

- Switch to the Azure Portal.
- Navigate to the **python-appinsights-SUFFIX** function app.
- Under **Functions**, select **Functions**, you should see the two python function apps deployed.

> **NOTE** Azure Function runtimes support various Python versions depending on the version selected.  See [Languages by runtime version](https://docs.microsoft.com/en-us/azure/azure-functions/supported-languages#languages-by-runtime-version).

- Review the function app code.  Notice the usage of the a storage account and a call to the database.

## 4 - Review Dependency Log Data

- Navigate to the **python-appinsights-SUFFIX** Application Insights resource, under **Monitoring**, select **Logs**
- Run the following query:

```kusto
dependencies
```

> - You should see the dependencies displayed for the calls to MySQL and Azure storage displayed.

Reviewer: Maybe include an image of the telemetry that is expected?

Author: Added.
